types/validation.go

```diff
@@ -273,11 +284,19 @@ func verifyCommitBatch(
 		// Validate signature.
 		voteSignBytes := commit.VoteSignBytes(chainID, int32(idx))
 
-		// add the key, sig and message to the verifier
-		if err := bv.Add(val.PubKey, voteSignBytes, commitSig.Signature); err != nil {
-			return err
+		cacheHit := false
+		if verifiedSignatureCache != nil {
+			cacheVal, sigIsInCache := verifiedSignatureCache[string(commitSig.Signature)]
+			cacheHit = sigIsInCache && bytes.Equal(cacheVal.ValidatorAddress, val.PubKey.Address()) && bytes.Equal(cacheVal.VoteSignBytes, voteSignBytes)
```

Comment:
this now assumes that the `val.PubKey` is uniquely identified by the `val.PubKey.Address()`, is that a safe assumption?
if it is not true, we will get false cache hits and will skip doing signature verification when we actually need to

Reply:
Can we store the validator's pubkey instead of the address to avoid making that assumption?

Reply:
Since validator pubkey is type `crypto.PubKey`, it seems we can't in general do an equality check with it, so we can't determine if two pubkeys match when checking for a cache hit. For example, `ed25519.PubKey` is uncomparable.
I've switched this to use the pubkey bytes, which might be better than address? but we can discuss further.